src/tc/actions/action.rs

```diff
@@ -49,62 +56,53 @@ impl Nla for TcAction {
+        // attributes.
+        // Messages of this type which do not specify [`Kind`], or which specify
+        // `Kind` more than once are malformed and should be rejected.
+        // We cannot ensure that `Kind` will be the first attribute in the
+        // `attributes` `Vec` (although it usually is).
+        // As a result, we need to determine `Kind` first, then parse the rest
+        // of the attributes.
+        let kind = match NlasIterator::new(buf.value())
+            .filter_map(|nla| {
+                let nla = match nla {
+                    Ok(nla) => nla,
+                    Err(e) => {
+                        return Some(
+                            Err(e).context("failed to parse action nla"),
                         )
                     }
-                    TcActionAttribute::Options(nlas)
+                };
+                match nla.kind() {
+                    TCA_ACT_KIND => Some(
+                        parse_string(nla.value())
+                            .context("failed to parse TCA_ACT_KIND"),
+                    ),
+                    _ => None,
                 }
-                TCA_ACT_INDEX => TcActionAttribute::Index(
-                    parse_u32(payload)
-                        .context("failed to parse TCA_ACT_INDEX")?,
-                ),
-                TCA_ACT_STATS => {
-                    let mut nlas = vec![];
-                    for nla in NlasIterator::new(payload) {
-                        let nla = nla.context("invalid TCA_ACT_STATS")?;
-                        nlas.push(
-                            TcStats2::parse_with_param(&nla, &kind).context(
-                                format!(
-                                    "failed to parse TCA_ACT_STATS for \
-                                    kind {kind}",
-                                ),
-                            )?,
-                        );
-                    }
-                    TcActionAttribute::Stats(nlas)
+            })
+            .collect::<Result<Vec<_>, _>>()
+        {
+            Ok(kinds) => {
+                if kinds.is_empty() {
+                    return Err(DecodeError::from("Missing TCA_ACT_KIND"));
                 }
-                TCA_ACT_COOKIE => TcActionAttribute::Cookie(payload.to_vec()),
-                TCA_ACT_IN_HW_COUNT => TcActionAttribute::InHwCount(
-                    parse_u32(payload)
-                        .context("failed to parse TCA_ACT_IN_HW_COUNT")?,
-                ),
-                _ => TcActionAttribute::Other(
-                    DefaultNla::parse(&buf)
-                        .context("failed to parse action nla")?,
-                ),
-            });
-        }
+                if kinds.len() > 1 {
+                    return Err(DecodeError::from("Duplicate TCA_ACT_KIND"));
+                }
+                kinds[0].clone()
+            }
+            Err(e) => return Err(DecodeError::from(e.to_string())),
```

Comment:
Does it affect unsupported cases?

Reply:
I am not 100% sure what you mean.

That said, I think this bit of code is kinda messy.  I will think about a better and more clear way to express this logic.

Reply:
Sorry, my mistake. 
Due to the diff is not match to the source code. The souce of parse move to `TcActionAttribute::parse_with_param`.

Reply:
Ok no worries.

Mark resolved?

Reply:
LGTM